spec/inputs/redis_spec.rb

```diff
@@ -61,3 +72,183 @@ def process(conf, event_count)
     process(conf, event_count)
   end
 end
+
+# unit tests ---------------------
+
+describe Logstash::Inputs::Redis do
+  let(:redis) { double('redis') }
+  let(:builder) { ->{ redis } }
+  let(:connection) { double('redis_connection') }
+  let(:connected) { [true] }
+  let(:data_type) { 'list' }
+  let(:cfg) { {'key' => 'foo', 'data_type' => data_type} }
+  let(:quit_calls) { [:quit] }
+
+  subject do
+    described_class.new(cfg).add_external_redis_builder(builder)
+  end
+
+  context 'construction' do
+    it 'registers the input' do
+      expect {subject.register}.not_to raise_error
+    end
+  end
+
+
+  context 'runtime for list data_type' do
+    before do
+      subject.register
+    end
+
+    context 'teardown when redis is unset' do
+      let(:quit_calls) { [:quit, :unsubscribe, :punsubscribe, :connection, :disconnect!] }
+
+      it 'does not attempt to quit' do
+        allow(redis).to receive(:nil?).and_return(true)
+        quit_calls.each do |call|
+          expect(redis).not_to receive(call)
+        end
+        expect {subject.teardown}.not_to raise_error
+      end
+    end
+
+    let(:accumulator) { [] }
+
+    it 'calling the run method, adds events to the queue' do
+      expect(redis).to receive(:blpop).at_least(:once).and_return(['foo', 'l1'])
+
+      allow(redis).to receive(:connected?).and_return(connected.last)
+      allow(redis).to receive(:quit)
+
+      tt = Thread.new do
+        sleep 0.01
+        subject.teardown
+      end
+
+      subject.run(accumulator)
+
+      tt.join
+
+      expect(accumulator.size).to be > 0
+    end
+
+    it 'multiple teardown calls, calls to redis once' do
+      subject.use_redis(redis)
+      allow(redis).to receive(:blpop).and_return(['foo', 'l1'])
+      expect(redis).to receive(:connected?).and_return(connected.last)
+      quit_calls.each do |call|
+        expect(redis).to receive(call).at_most(:once)
+      end
+
+      subject.teardown
+      connected.push(false) #can't use let block here so push to array
+      expect {subject.teardown}.not_to raise_error
+      subject.teardown
+    end
+  end
+
+  context 'for the subscribe data_types' do
+    def run_it_thread(inst)
```

Comment:
Hi, 
  but why not adding a module, and then make it an [RSpec helper](https://www.relishapp.com/rspec/rspec-core/docs/helper-methods/define-helper-methods-in-a-module#include-a-module-in-all-example-groups). This would still keep the visibility and also have this methods located their own container/class/module. 

But I'm open to education of the benefits/cons of having this here like this. Is my that have had always the taste of having this in their own helpers.